Vertical scrolled plans for mobile A/B test

ddc22 opened the pull request:
#### Changes proposed in this Pull Request

* Based on project thread pcbrnV-XN-p2
* Provides a vertically scrollable card view for mobile resolutions in the signup flow
* Add a "device" prop to events related funnel to be analysed which can have values ["DESKTOP", "MOBILE", "TABLET"]

| Mobile            | IPAD               |
| ------------- | ------------- |
| ![image](https://user-images.githubusercontent.com/3422709/104597435-558aa000-569b-11eb-95d4-289141402932.png)  | ![image](https://user-images.githubusercontent.com/3422709/104597533-7521c880-569b-11eb-8ef8-49f21ec31b56.png)  |


### Testing instructions

- ~~Hardcode experiment to be activated set true for `isInVerticalScrollingPlansExperiment`~~
~~ https://github.com/Automattic/wp-calypso/pull/48728/files#diff-f4523a33abda8414624dfd8061cffa9cc3b52e263f9a755706831b28b7b55cecR324-R325 ~~

- Activate the experiment that has been setup on ExPlat. Check pcbrnV-XN-p2 for link.
- Login to wordpress 
- In one tab Goto  /start and go to the plans step (Sign up flow plans step)
- In another go to the home page and go to the plan page
<img src="https://user-images.githubusercontent.com/3422709/104598724-eada6400-569c-11eb-8317-73c5f7b7a460.png" width="400" >

- Switch to mobile resolutions in both tabs and check whether the styles match across both mobile views (Width < 1045px)
- Make sure the sign up flow plans page in the desktop resolution remains unchanged in caparison with prod
- Make sure the logged in plans page remains unchanged in caparison with prod

#### Test to make sure control remains unchanged
- Activate control in experiment
- Make sure the signup flow plans page as well as the logged in plan page match production exactly

ddc22 committed:
Added listener to listen on resolution on activate a flag which switches between the old mobile card view and table view of plans.
Anything below < 1040px will have the vertical scroll plans
Anything above that will have the traditional table view of plans
A tracks props device type was added on each submission of the tracks event

ddc22 committed:
Refactor viewport resolution and add prop to tracks event

ddc22 committed:
Activate experiment

ddc22 committed:
Fix test failures

southp commented on packages/viewport/src/index.js:
```diff
@@ -187,3 +187,40 @@ export function subscribeIsDesktop( listener ) {
+
+/******************************************/
+/*       Vertical Scroll Experiment       */
+/******************************************/
```
Should we annotate this here? An A/B test is an application-side logic, but `@automattic/viewport` is a general package that only . Putting it another way: would these three newly-added functions become useless outside of the vertical scroll experiment?

ddc22 replied:
So this is something that I would attribute to a cleanup first strategy, So in the event of a cleanup this would be easier and it serves as a boundary in where nobody should touch this code because it most probably will be refactored out.

southp commented on packages/viewport/src/index.js:
```diff
@@ -187,3 +187,40 @@ export function subscribeIsDesktop( listener ) {
+/*       Vertical Scroll Experiment       */
+/******************************************/
+
+//TODO: To be refactored using above constants and helper methods
```
I am not sure I understand this. Could you clarify which constants and which helper methods? :)

ddc22 replied:
I have updated the comment to be more clear

southp commented on packages/viewport/src/index.js:
```diff
@@ -187,3 +187,40 @@ export function subscribeIsDesktop( listener ) {
+}
+
+export function isMobileResolution() {
+	if ( typeof window === 'object' ) {
```
To be consistent with the other functions, `! isServer` could be better.

ddc22 replied:
Done

southp commented on packages/viewport/src/index.js:
```diff
@@ -187,3 +187,40 @@ export function subscribeIsDesktop( listener ) {
+}
+
+export function isTabletResolution() {
+	if ( typeof window === 'object' ) {
```
The same as the above.

ddc22 replied:
Done

southp commented on client/state/analytics/actions/record.js:
```diff
@@ -18,8 +19,14 @@ export const recordEvent = ( service, args ) => ( {
+		...properties,
+		device_type: resolveDeviceTypeByViewPort(),
+	};
+	return recordEvent( 'tracks', { name, properties: finalProperties } );
```
This is huge and will affect all the events. Thus, even though it should work for what we want, I'd like us to not rush it but hear from the data team first. @aaronyan @withinboredom could you pitch in here? 

@jdc91 In the future, I'd recommend to isolate this kind of change into a separate PR so we can have a clearer view to isolated pieces. In this case, this change together with the viewport package update can be extracted from the experiment itself. Of course, let's not do that now, just wanted to let you be aware of this strategy :)

ddc22 replied:
Thanks @southp I will keep that in mind.

ddc22 replied:
If you think this is a bit too big of a change, I can go over the exact event thats that needs this prop and introduce it. Emre has provided me with a set of fitting events to include, so this will be a less impacting change.

niranjan-uma-shankar replied:
>  I can go over the exact event thats that needs this prop and introduce it.

My two cents - I would second this approach. Since we can send properties as an argument to the `recordTracksEvent` function, it'll be safer to send the `device_type` prop only for the exposure event. Probably try to completely avoid touching this file, as any inadvertent breakage here could bring down all our tracking.

ddc22 replied:
Moved this to each sepcific event.

niranjan-uma-shankar reviewed (commented):
1. After selecting a plan, I'm unable to proceed to checkout and instead get redirected to a login page. I observe this if I'm assigned to either control or treatment and in a mobile viewport. I don't get this issue if I'm in desktop viewport, or if I don't assign myself to the experiment. Not sure if it's something specific to my setup though, could you please confirm?
2. For both [mobile](url) and [tablet](https://d.pr/i/zJo71h) views, the Free plan option appears in the plan grid in addition to the "Start with free" option in the header. I believe we shouldn't show the free option in the plans grid.
3. The experiment loading indicator is not aligned properly. I'm not sure if it's easy to see it in the screenshot below, but the pulsating dot indicator appears just above the word `that's` in the header. 

<img width="812" alt="Screenshot 2021-01-25 at 2 33 37 PM" src="https://user-images.githubusercontent.com/1269602/105690573-2b9a6e80-5f22-11eb-96f4-c6895cbcc6ae.png">

ddc22 committed:
Remove global changes to tracks events and target specific events
Revert changes to tests
Fixed loading indicator styles
Removed free plan showing in mobile resolutions

ddc22 commented:
@niranjan-uma-shankar Quoting your suggestions

> 1. After selecting a plan, I'm unable to proceed to checkout and instead get redirected to a login page. I observe this if I'm assigned to either control or treatment and in a mobile viewport. I don't get this issue if I'm in desktop viewport, or if I don't assign myself to the experiment. Not sure if it's something specific to my setup though, could you please confirm?
> 2. For both [mobile](url) and [tablet](https://d.pr/i/zJo71h) views, the Free plan option appears in the plan grid in addition to the "Start with free" option in the header. I believe we shouldn't show the free option in the plans grid.
> 3. The experiment loading indicator is not aligned properly. I'm not sure if it's easy to see it in the screenshot below, but the pulsating dot indicator appears just above the word `that's` in the header.

Fixed 2 and 3, could not reproduce 1 I will recheck to be sure.

----

I still could not recreate 1, Was anyone else able to recreate it?

ddc22 committed:
Add props to additional events when payment completed

niranjan-uma-shankar commented on packages/viewport/src/index.js:
```diff
@@ -189,3 +189,32 @@ export function subscribeIsDesktop( listener ) {
+	return false;
+}
+
+export const DEVICE_MOBILE = 'MOBILE';
```
Nitpick: It would be consistent with other analytics props if this value is in lowercase instead of all uppercase.

ddc22 replied:
Ah yes I will change this! 👍

niranjan-uma-shankar commented on client/lib/analytics/signup.js:
```diff
@@ -66,7 +67,8 @@ export function recordSignupComplete(
 		// Tracks
-		recordTracksEvent( 'calypso_new_user_site_creation', { flow } );
+		const device = resolveDeviceTypeByViewPort();
+		recordTracksEvent( 'calypso_new_user_site_creation', { flow, device } );
```
Nitpick: This is technically not device but viewport. I could still be on desktop and yet get the value "tablet".

ddc22 replied:
Yes @niranjan-uma-shankar this is a convention to match the future function which is resolveDeviceByUserAgent.

niranjan-uma-shankar reviewed (approved):
I'm still seeing the login redirect issue in Chrome, but not in FF. Maybe you might want to check if anyone else is seeing the same problem, if not then feel free to merge. 

A few nitpicky comments below but otherwise everything works well.

sirbrillig commented on packages/viewport/src/index.js:
```diff
@@ -189,3 +189,32 @@ export function subscribeIsDesktop( listener ) {
+export function resolveDeviceTypeByViewPort() {
+	let deviceType;
+	if ( isMobile() ) {
+		deviceType = DEVICE_MOBILE;
```
Really minor, but each of these could be an early return, which would remove the need for the `deviceType` variable as well as each `else` and might make the function slightly more elegant.

```js
if ( isMobile() ) {
  return DEVICE_MOBILE;
}
if ( isTabletResolution() ) {
  return DEVICE_TABLET;
}
return DEVICE_DESKTOP;
```

ddc22 replied:
Thanks @sirbrillig that is definitely more elegant 👍 ☺️

sirbrillig commented on client/my-sites/checkout/composite-checkout/hooks/use-create-payment-complete-callback.tsx:
```diff
@@ -336,13 +340,15 @@ function recordPaymentCompleteAnalytics( {
 			total: responseCart.total_cost_integer,
 			currency: responseCart.currency,
 			payment_method: wpcomPaymentMethod || '',
+			device,
```
This looks fine from a checkout perspective! 👍

ddc22 committed:
Minor fixes

ddc22 committed:
Scroll plans div only instead of entire viewport

southp reviewed (approved):
The new change LGTM 🚢
